integer.rb

```diff
@@ -55,6 +55,29 @@ def ord
     return self
   end
 
+  #  Document-method: Integer#times
+  #  call-seq:
+  #     int.times {|i| block }  ->  self
+  #     int.times               ->  an_enumerator
+  #
+  #  Iterates the given block +int+ times, passing in values from zero to
+  #  <code>int - 1</code>.
+  #
+  #  If no block is given, an Enumerator is returned instead.
+  #
+  #     5.times {|i| print i, " " }   #=> 0 1 2 3 4
+  def times
+    unless Primitive.block_given_p
+      return Primitive.cexpr! 'SIZED_ENUMERATOR(self, 0, 0, int_dotimes_size)'
+    end
```

Comment:
Would
```ruby
return to_enum(:times) { self < 0 ? 0 : self } unless block_given?
```
work fine here too?

Being able to use `block_given?` in core Ruby code seems nice.
Since `block_given?` is frequently used in user programs it should be well optimized for both cases.

Reply:
> Would return to_enum(:times) { self < 0 ? 0 : self } unless block_given? work fine here too?

Feature-wise, it would work. But the purpose of using `Primitive.` was to avoid degrading performance around these parts just for making it possible to inline a block call in JIT (in the future). IMO, a PR to allow JIT inlining and one to refactor the C calls can be separated.

> Being able to use block_given? in core Ruby code seems nice.
> Since block_given? is frequently used in user programs it should be well optimized for both cases.

True. The main concern here is the overhead of `block_given?` method call which doesn't exist in `Primitive.block_given_p`, so I think we might be able to optimize VM in some ways like:

* Add `opt_block_given_p` insn.
  * This would increase the VM code size and pressure i-cache, so I don't like this idea.
* Annotate `block_given?` itself with 'inline', and eliminate the method call overhead when a method has the annotation.

Either way, I think it should be done in a different PR.

Reply:
I see, thanks for the details.
FWIW, TruflfeRuby AST-inlines `block_given?` (and rewrites it to a call if it would end up calling a non-standard definition of it), so it's similar to having a `opt_block_given_p` insn in MRI.

> * Annotate `block_given?` itself with 'inline', and eliminate the method call overhead when a method has the annotation.

Would that also work without the JIT? I guess it would be a new optimization for core library Ruby code, that applies both with/without `--jit`?

Reply:
> Would that also work without the JIT?

No. On the interpreter, calling a builtin method that is implemented using only `Primitive` is only slightly faster than calling a regular C method, whether it's `Primitive.attr! 'inline'` or not. Without JIT, calling `block_given?` is guaranteed to be slower than `Primitive.block_given_p` unless we introduce `opt_block_given_p` insn.

Introducing `opt_block_given_p` could actually be the right solution if it's a repeated pattern and also gives a speedup in benchmarks. It might be worth a try (again, a scope for another PR though).